include/seqan3/range/view/detail.hpp

```diff
@@ -13,301 +13,296 @@
+        return combined_adaptor{static_cast<derived_type &&>(me), std::forward<arg_t>(arg)};
     }
 
+    //!\overload
+    template <typename arg_t>
+    constexpr friend auto operator|(adaptor_base const & me, arg_t && arg)
+    {
+        return combined_adaptor{static_cast<derived_type const &>(me), std::forward<arg_t>(arg)};
+    }
+};
+
+// ============================================================================
+//  combined_adaptor
+// ============================================================================
+
+/*!\brief Template for range adaptor closure objects that consist of two other range adaptor closure objects.
+ * \tparam left_adaptor_t  Type of the first stored adaptor.
+ * \tparam right_adaptor_t Type of the second stored adaptor.
+ * \ingroup view
+ *
+ * \details
+ *
+ * If invoked with a range, this adaptor resolves to piping the range into left adaptor and the resulting
+ * range into the right adaptor.
+ */
+template <typename left_adaptor_t, typename right_adaptor_t>
+class combined_adaptor :
+    public adaptor_base<combined_adaptor<left_adaptor_t, right_adaptor_t>,
+                        left_adaptor_t,
+                        right_adaptor_t>
+{
 private:
-    /*!\name Constructors, destructor and assignment
-     * \brief All are declared private to prevent direct use of the CRTP base.
-     * \sa https://isocpp.org/blog/2017/04/quick-q-prevent-user-from-derive-from-incorrect-crtp-base
-     * \{
-     */
-    pipable_adaptor_base() = default;
-    constexpr pipable_adaptor_base(pipable_adaptor_base const &) = default;
-    constexpr pipable_adaptor_base(pipable_adaptor_base &&) = default;
-    constexpr pipable_adaptor_base & operator =(pipable_adaptor_base const &) = default;
-    constexpr pipable_adaptor_base & operator =(pipable_adaptor_base &&) = default;
-    ~pipable_adaptor_base() = default;
-    //!\brief befriend the derived type so that it can instantiate
-    friend derived_type;
-    //!\}
+    //!\brief Type of the CRTP-base.
+    using base_type = adaptor_base<combined_adaptor<left_adaptor_t, right_adaptor_t>,
+                                   left_adaptor_t,
+                                   right_adaptor_t>;
+
+    //!\brief Befriend the base class so it can call impl().
+    friend base_type;
+
+    //!\brief Combine all arguments via `operator|`.
+    template <std::ranges::InputRange urng_t,
+              typename left_adaptor_t_,
+              typename right_adaptor_t_>
+    static auto impl(urng_t && urange, left_adaptor_t_ && left_adaptor, right_adaptor_t_ && right_adaptor)
+    {
+        return std::forward<urng_t>(urange)
+             | std::forward<left_adaptor_t_>(left_adaptor)
+             | std::forward<right_adaptor_t_>(right_adaptor);
+    }
+
+public:
+    //!\brief Inherit the base type's constructors.
+    using base_type::base_type;
+
+    //!\brief Store both arguments in the adaptor.
+    combined_adaptor(left_adaptor_t l, right_adaptor_t r) :
```

Comment:
You copy here, why do you forward `l` and `r` in the member initialization?

Reply:
does `std::forward` move a copy automatically? I thought this only works if you pass the parameter by `&&`

Reply:
What do you mean by moving a copy? The only thing that `std::forward<T>(t)` does is `static_cast<T&&>(t) ` that's the full implementation.

It has no knowledge whatsoever about how your function took any of it's parameters, it just makes use of reference collapsing so that if `T` was e.g. `int &` than the cast will not change the type. But if the type was `int` it will become `int &&`.

That's exactly the behaviour we want in that situation or not?

Reply:
your explanation above resolved my misunderstanding!